# Exploratory Data Analysis

In [1]:
data_dir = '../data/raw/'

In [6]:
import pandas as pd


column_names = ['user_id', 'item_id', 'rating', 'timestamp']

df = pd.read_csv(
    data_dir + 'u.data',
    sep='\t',
    names=column_names,
    dtype={'user_id': int, 'item_id': int, 'rating': int},
    parse_dates=['timestamp']
)

df.head(), df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   user_id    100000 non-null  int32 
 1   item_id    100000 non-null  int32 
 2   rating     100000 non-null  int32 
 3   timestamp  100000 non-null  object
dtypes: int32(3), object(1)
memory usage: 1.9+ MB


C:\Users\newro\AppData\Local\Temp\ipykernel_9448\2485261898.py:6: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(


(   user_id  item_id  rating  timestamp
 0      196      242       3  881250949
 1      186      302       3  891717742
 2       22      377       1  878887116
 3      244       51       2  880606923
 4      166      346       1  886397596,
 None)

In [7]:
import pandas as pd


# Set the file paths for each data file
u_data_path = data_dir + 'u.data'
u_info_path = data_dir + 'u.info'
u_item_path = data_dir + 'u.item'
u_genre_path = data_dir + 'u.genre'
u_user_path = data_dir + 'u.user'
u_occupation_path = data_dir + 'u.occupation'

# Read data files into DataFrames
u_data = pd.read_csv(u_data_path, sep='\t', names=['user_id', 'item_id', 'rating', 'timestamp'])
u_info = pd.read_csv(u_info_path, header=None, names=['info'])
u_item = pd.read_csv(u_item_path, sep='|', encoding='latin-1', names=['movie_id', 'movie_title', 'release_date', 'video_release_date', 'IMDB_URL', 'unknown', 'Action', 'Adventure', 'Animation', 'Children', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western'])
u_genre = pd.read_csv(u_genre_path, sep='|', names=['genre'])
u_user = pd.read_csv(u_user_path, sep='|', names=['user_id', 'age', 'gender', 'occupation', 'zip_code'])
u_occupation = pd.read_csv(u_occupation_path, names=['occupation'])

# Display basic insights for each dataset
print("u.data:")
print(u_data.head())
print("\nDescriptive statistics for u.data:")
print(u_data.describe())
print("\nu.info:")
print(u_info)
print("\nu.item:")
print(u_item.head())
print("\nDescriptive statistics for u.item:")
print(u_item.describe())
print("\nu.genre:")
print(u_genre)
print("\nu.user:")
print(u_user.head())
print("\nDescriptive statistics for u.user:")
print(u_user.describe())
print("\nu.occupation:")
print(u_occupation)

u.data:
   user_id  item_id  rating  timestamp
0      196      242       3  881250949
1      186      302       3  891717742
2       22      377       1  878887116
3      244       51       2  880606923
4      166      346       1  886397596

Descriptive statistics for u.data:
            user_id        item_id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

u.info:
             info
0       943 users
1      1682 items
2  100000 ratings

u.item:
   movie_id        movie_title release_date  video_relea

In [11]:
male_count = u_user[u_user['gender'] == 'M'].shape[0]
total_users = u_user.shape[0]

(male_count / total_users) * 100

71.04984093319194

In [21]:
occupations = u_user.occupation.unique()
total_users = u_user.shape[0]

for o in occupations:
    o_count = u_user[u_user.occupation == o].shape[0]
    print(f'{(o_count / total_users) * 100:.2f}%\t{o}')

2.86%	technician
11.13%	other
4.77%	writer
3.39%	executive
8.38%	administrator
20.78%	student
1.27%	lawyer
10.07%	educator
3.29%	scientist
1.91%	entertainment
7.00%	programmer
5.41%	librarian
0.74%	homemaker
2.97%	artist
7.10%	engineer
2.76%	marketing
0.95%	none
1.70%	healthcare
1.48%	retired
1.27%	salesman
0.74%	doctor


## Insights

- Average rating score is 3.52
- Top 1 genre is Comedy (30% of movies)
- Least popular genre is Fantasy (1.3% of movies)
- Average user age is 34 years old (ages from 7 to 73 years old are in the dataset)
- Ratio of user gender is 71% / 29% for Male / Female
- Student is the most popular occupation (20.7% of users)
- Homemaker and Docter are least popular occupations (0.74% of users)